File: Common_CF.m
classdef Common_CF
    %COMMON_CF  Class containing fundamental physical constants and utility functions
    %
    %   This class defines physical constants in SI and astronomical units, and provides
    %   methods for cosmological calculations. All constants are defined as read-only
    %   properties with fixed units as noted below.
    %
    %   Example:
    %     cf = Common_CF;
    %     lumDist = cf.distance(0.5);  % Calculate luminosity distance at z=0.5
    %
    % Version:
    %   1.0 (2025-03-20) Initial implementation
    %   1.1 (2025-04-15) Added cosmological parameter customization
    
    properties (Constant = true)
        %MS Solar mass [kg]
        %   Reference: IAU 2015 Resolution B3
        Ms = 1.9892e30;
        
        %H0 Hubble constant [s^-1]
        %   Current value: 72 km/s/Mpc converted to inverse seconds
        H0 = 72/(3.0856776e19);
        
        %PC Parsec [m]
        %   Exact value per IAU 2015 Resolution B2
        pc = 3.0856776e16;
        
        %GPC Gigaparsec [m]
        Gpc = 3.0856776e25;
        
        %T1YEAR Tropical year [s]
        %   Mean tropical year duration at J2000.0
        T1year = 3.1556925e7;
        
        %R Schwarzschild radius multiplier
        %   Used for defining integration boundaries (default 499)
        R = 499;
        
        %C Speed of light in vacuum [m/s]
        %   Exact value per SI definition
        c = 299792458;
        
        %G Gravitational constant [m^3 kg^-1 s^-2]
        %   CODATA 2018 recommended value
        G = 6.67430e-11;
        
        %AU Astronomical unit [m]
        %   IAU 2012 exact definition
        AU = 149597870700;
    end
    
    methods
        function MK = readinput(obj, filename)
            %READINPUT Read parameter matrix from CSV file
            %
            %   MK = READINPUT(filename) loads a comma-delimited matrix from the
            %   specified file. The file should contain numerical values only.
            %
            %   Input:
            %       filename    Path to input file (string)
            %
            %   Output:
            %       MK          Parameter matrix [N×M double]
            
            MK = readmatrix(filename, 'Delimiter', ',');
        end
        
        function ret = distance(obj, z)
            %DISTANCE Calculate luminosity distance for given redshift
            %
            %   D = DISTANCE(z) computes the EM luminosity distance using
            %   standard ΛCDM cosmology with default parameters:
            %     - H0 = 72 km/s/Mpc
            %     - Ωm = 0.29
            %     - ΩΛ = 0.71
            %
            %   Reference:
            %     Bennett et al. 2003, ApJS 148, 1 (WMAP first-year results)
            %
            %   Input:
            %       z       Redshift [dimensionless]
            %
            %   Output:
            %       dL      Luminosity distance [m]
            
            % Cosmological parameters (WMAP9 values)
            O_M = 0.29;     % Matter density parameter
            O_A = 0.71;     % Dark energy density parameter
            
            % Numerical integration setup
            zz = linspace(0, z, 50);    % Integration points
            dzz = zz(3) - zz(2);        % Step size
            
            % Integrate comoving distance (Eq. 18 in Hogg 1999)
            ret = obj.c/obj.H0 * (1+z) * ...
                  sum(1./sqrt(O_M*(1+zz).^3 + O_A)) * dzz;
        end
    end
end

File: src_python/common_cf.py
# common_cf.py
import numpy as np
from scipy import constants

# ==================================
# 1. 物理常数 (Physical Constants)
# ==================================
c = constants.c          # 光速 (m/s)
G = constants.G          # 万有引力常数 (N m^2 / kg^2)
M_sun = 1.989e30         # 太阳质量 (kg)
Mpc = 1e6 * constants.parsec # 百万秒差距 (m)

# =======================================
# 2. 宇宙学参数 (Cosmological Parameters)
# (Based on Planck 2018)
# =======================================
H0 = 67.4                 # 哈勃常数 (km/s/Mpc)
h = H0 / 100              # 无量纲哈勃常数

Omega_m = 0.315           # 物质密度参数
Omega_lambda = 0.685      # 暗能量密度参数

# 这个是被遗漏的常数
Omega_k = 1.0 - Omega_m - Omega_lambda  # 曲率密度参数 (在此模型中为 0)

# =======================================
# 3. 核心函数 (Core Function)
# =======================================
def distance(z):
    """
    根据红移z计算光度距离 (单位: m)
    使用数值积分计算。
    """
    from scipy.integrate import quad
    
    def integrand(z_prime):
        # 完整的积分项，包含了曲率项
        return 1.0 / np.sqrt(Omega_m * (1 + z_prime)**3 + Omega_k * (1 + z_prime)**2 + Omega_lambda)
        
    # 从0到z积分
    integral, _ = quad(integrand, 0, z)
    
    # 同移距离 (Comoving Distance)
    # H0_in_SI 是以 s^-1 为单位的哈勃常数
    H0_in_SI = (H0 * 1000) / Mpc
    d_C = (c / H0_in_SI) * integral
    
    # 光度距离 (Luminosity Distance)
    # d_L = (1+z) * d_C  (在平坦宇宙中)
    # 更通用的公式涉及到 d_M (横向同移距离)，但在平坦宇宙中 d_M = d_C
    d_L = (1 + z) * d_C
    
    return d_L

if __name__ == '__main__':
    # 测试
    print(f"定义的宇宙学参数:")
    print(f"  Omega_m = {Omega_m}")
    print(f"  Omega_lambda = {Omega_lambda}")
    print(f"  Omega_k = {Omega_k} (计算得出)")

    z_test = 0.1
    d_L_meters = distance(z_test)
    d_L_Mpc = d_L_meters / Mpc
    print(f"\n红移 z = {z_test}对应的光度距离是: {d_L_Mpc:.4f} Mpc")